### PR TITLE
Issue 72 & 73 configurability fixes

### LINKED ISSUE #72: Consul Connect Upstreams should be configurable
## Feature description
Add an input variable and corresponding template stanza to enable configurable upstreams
## Why is it needed?
Support outgoing integrations from MinIO over Consul Connect
## Definition of done
Nomad job template renders a configurable and optional list of upstreams 
## Checklist (after created issue)
- [x] Added label(s)
- [x] Added to project
- [x] Added to milestone

### LINKED ISSUE #73: Add optional extra commands to MC job
## Feature description
Expose an input variable with a list of additional commands to run with MC container
## Why is it needed?
This allows post-deploy customizations using MC to occur without code changes to MinIO module
## Definition of done
MC runs the list of commands provided in input variable, if provided.
## Checklist (after created issue)
- [x] Added label(s)
- [x] Added to project
- [x] Added to milestone

### PR DESCRIPTION
**Pull request by oschistad**
## Closes/fixes/resolves issue(s)?
#72 #73 
## What was added/changed/fixed?
Input variables minio_upstreams, mc_extra_commands
## Checklist (after created PR)
- [x] Added reviewers
- [x] Added assignee(s)
- [x] Added label(s)
- [x] Added to project
- [x] Added to milestone

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -47,7 +47,7 @@ data "template_file" "nomad_job_minio" {
     envs            = local.minio_env_vars
     use_host_volume = var.use_host_volume
     use_canary      = var.use_canary
-
+    upstreams = jsonencode(var.minio_upstreams)
```
Comment (pdmthorsrud) [inline main.tf]:
```suggestion
    upstreams       = jsonencode(var.minio_upstreams)
```
I think I got the amount spaces right 😅

File: conf/nomad/minio.hcl [hunk 2]
```diff
@@ -52,7 +52,16 @@ job "${service_name}" {
...
-        sidecar_service {}
+        sidecar_service {
+ %{ for upstream in jsondecode(upstreams) }
+          proxy {
+            upstreams {
+              destination_name = "${upstream.service_name}"
+              local_bind_port = "${upstream.port}"
+            }
+          }
+%{ endfor }
```
Comment (pdmthorsrud) [inline conf/nomad/minio.hcl]: It's only upstreams stanza that needs to be looped, isn't it? https://github.com/fredrikhgrelland/terraform-nomad-presto/blob/c1a54347b1689a6234067db2245e8a1ff71697fe/conf/nomad/presto.hcl#L33-L49
```suggestion
          proxy {
 %{ for upstream in jsondecode(upstreams) }
            upstreams {
              destination_name = "${upstream.service_name}"
              local_bind_port = "${upstream.port}"
            }
%{ endfor }
          }
```

Comment (oschistad) [reply conf/nomad/minio.hcl]: Yes, good catch. Ty!

**Review by pdmthorsrud — DISMISSED** [review]
LGTM, thank you! It's really cool to see people contributing!

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,8 @@
 ### Added
 - Added canary deployment and switch #64
 - Added intentions sections to README.md #65
+- Added input variable and template stanza for Consul Connect upstreams
+- Added input variable for MC extra commands, appended to command in MC job
```
Comment (claesgill) [inline CHANGELOG.md]:
```suggestion
- Added input variable and template stanza for Consul Connect upstreams #72
- Added input variable for MC extra commands, appended to command in MC job #73
```

File: conf/nomad/minio.hcl [hunk 4]
```diff
@@ -52,7 +52,16 @@ job "${service_name}" {
         timeout   = "4s"
       }
       connect {
-        sidecar_service {}
+        sidecar_service {
+          proxy {
+ %{ for upstream in jsondecode(upstreams) }
+            upstreams {
+              destination_name = "${upstream.service_name}"
+              local_bind_port = "${upstream.port}"
```
Comment (claesgill) [inline conf/nomad/minio.hcl]:
```suggestion
              local_bind_port  = "${upstream.port}"
```

File: variables.tf [hunk 5]
```diff
@@ -97,3 +97,19 @@ variable "buckets" {
...
+    port         = number,
+  }))
+  description = "List of upstream services (list of object with service_name, port)"
+  default = []
+}
+
+variable "mc_extra_commands" {
+  type = list(string)
+  default = [""]
+  description = "Extra commands to run in MC container after creating buckets"
```
Comment (claesgill) [inline variables.tf]:
```suggestion
  type        = list(string)
  default     = [""]
  description = "Extra commands to run in MC container after creating buckets"
```

File: variables.tf [hunk 6]
```diff
@@ -97,3 +97,19 @@ variable "buckets" {
...
   default     = []
 }
+
+variable "minio_upstreams" {
+  type = list(object({
+    service_name = string,
+    port         = number,
+  }))
+  description = "List of upstream services (list of object with service_name, port)"
+  default = []
```
Comment (claesgill) [inline variables.tf]:
```suggestion
  description = "List of upstream services (list of object with service_name, port)"
  default     = []
```

**Review by claesgill — CHANGES_REQUESTED** [review]
PR looks good :rocket: 

Only have some minor comments, and I'd like you to update the *Inputs* section in the readme with the new variables that was added.

**Review by zhenik — COMMENTED** [review]
PR seems good, could you also update README.md, sections: example, inputs?

**Review by zhenik — DISMISSED** [review]
LGTM

**Review by pdmthorsrud — DISMISSED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (claesgill) [reply CHANGELOG.md]: Please merge this 😄

**Review by zhenik — APPROVED** [review]

**Review by claesgill — APPROVED** [review]
# 🚀

**Pull request merged**